src/gradients.rs

```diff
@@ -248,18 +245,43 @@ pub trait GradientProvider {
 ///
 /// Most implementations of this trait will have sub structs that also
 /// implement [CanUpdateWithGradients].
-///
-/// For example the [Linear] model just calls update on its weight & bias:
-/// ```ignore
-/// impl<const I: usize, const O: usize> CanUpdateWithGradients for Linear<I, O> {
-///     fn update<G: GradientProvider>(&mut self, grads: &mut G) {
-///         self.weight.update(grads);
-///         self.bias.update(grads);
-///     }
-/// }
-/// ```
 pub trait CanUpdateWithGradients {
-    fn update<G: GradientProvider>(&mut self, grads: &mut G);
+    /// Updates self given the [GradientProvider]. When any amount of parameters that
+    /// should be updated are NOT present in `G`, then this function should
+    /// populate the resulting [MissingGradients] struct with the names.
+    fn update<G: GradientProvider>(&mut self, grads: &mut G, missing: &mut MissingGradients);
+}
+
+/// An struct that holds names of parameters that were missing associated gradients.
+#[derive(Debug, Default)]
+pub struct MissingGradients {
```

Comment:
rename to something like:
- `TensorsMissingGradients`
- `TensorGradientsMissing`
- `MissingTensorGradients`
- `MissingTensors`
- `UnusedParams`
- `ParamsMissingGradients`
- `NotUpdated`
- `TensorsNotUpdated`
- `ParamsNotUpdated`
- `NonUpdatedParams`